File: solutions/java/code-carl/ch09_dp/T21_LC0337_Rob.java
/**
 * Practice template for:
 * - LeetCode 337. House Robber III
 *   https://leetcode.cn/problems/house-robber-iii/
 */
/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2024/11/23 19:22
 * @comment
 */
public class T21_LC0337_Rob {
    public int rob(TreeNode root) {
        int[] res = traversal(root);
        return Math.max(res[0], res[1]);
    }

    public int[] traversal(TreeNode root) {
        if (root == null) {
            return new int[2];
        }
        // 0 no
        // 1 do
        int[] left = traversal(root.left);
        int[] right = traversal(root.right);
        int[] res = new int[2];
        res[0] = Math.max(left[0], left[1]) + Math.max(right[0], right[1]);
        res[1] = root.val + left[0] + right[0];
        return res;
    }


    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}

File: templates/code-carl/ch09_dp/t21_lc0337_rob.py
"""LeetCode 337. House Robber III
https://leetcode.cn/problems/house-robber-iii/
"""

from __future__ import annotations

from typing import Optional

from template_py.ch06_tree.tree_node import TreeNode


class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        pass

if __name__ == "__main__":
    def build_tree(nodes):
        if not nodes:
            return None
        values = list(nodes)
        root_val = values.pop(0)
        if root_val is None:
            return None
        root = TreeNode(root_val)
        queue = [root]
        while values and queue:
            node = queue.pop(0)
            if values:
                left_val = values.pop(0)
                if left_val is not None:
                    node.left = TreeNode(left_val)
                    queue.append(node.left)
            if values:
                right_val = values.pop(0)
                if right_val is not None:
                    node.right = TreeNode(right_val)
                    queue.append(node.right)
        return root

    solution = Solution()
    assert solution.rob(build_tree([3, 2, 3, None, 3, None, 1])) == 7
    assert solution.rob(build_tree([3, 4, 5, 1, 3, None, 1])) == 9
    print("All tests passed.")
